[Feature-14832][Listener]Implementation of Listener Mechanism

weixiaonan1 opened the pull request:
This closes https://github.com/apache/dolphinscheduler/issues/14832

## Purpose of the pull request

This pull request adds a new type of alert plugin instance: global alert plugin instance.  The global alert plugin instance will receive and post all system-generated events by default. Now support 10 types of events.

## Brief change log
1. add a global alert group (id: 2) which users cannot modify/delete, the global alert instance will be added into/ removed from this group automatically.
![image](https://github.com/apache/dolphinscheduler/assets/33857431/f9c7794b-4136-4dca-995d-97715ed355fe)

2. In addition to the original alarm instance type, add a new option for global alarm instances.
![image](https://github.com/apache/dolphinscheduler/assets/33857431/2e113ab6-f43e-4abb-924c-032251ac47d0)

3. add 10 types of events: `ServerDownListenerEvent`, `ProcessDefinitionCreatedLitenerEvent`, `ProcessDefinitionUpdatedLitenerEvent`, `ProcessDefinitionDeletedLitenerEvent`, `ProcessStartListenerEvent`, `ProcessEndListenerEvent`, `ProcessFailListenerEvent`, `TaskStartListenerEvent`, `TaskEndListenerEvent`, `TaskFailListenerEvent`. The event will be saved in table `t_ds_listener_event`.
4. In addition to `AlertBootstrapService`, add `ListenerEventPostService` in `AlertServer`: get pending listener events from `t_ds_listener_event`, get global alert instances and post these events.

### database change
1. insert global alert group into `t_ds_alertgroup`
2. add `instance_type`, `warning_type` in `t_ds_alert_plugin_instance`. `alert instance type`: 0 normal; 1 global. `warning_type`: warning type of normal alert instance which existed in `plugin_instance_params` previously.
3. new table `t_ds_listener_event`: save global listener events, which is simpler than `t_ds_alert`.

## Brief change log

weixiaonan1 committed:
first commit

weixiaonan1 commented:
Due to OSPP time constraints, some improments need to be done in future PRs. 

1. Refine global alert instances: when creating global alert instance, we can choose event types to subscribe to. The current global alert instance subscribes to all events by default. If we can choose event types,  there is no longer a need for the `default admin warning group`. We can create a global alert instance that only subscribe to `ServerDownListenerEvent`to replace the `default admin warning group`.
2. Is it necessary to choose `warning type` for normal alert instances? In order to be compatible with the logic of normal alert instances, we have reserved the option of warning type (success, fail, all). Warning type is actually a collection of event types: `success` is a collection of ProcessEndEvent; `fail` is a collection of ProcessFailEvent, ProcessTimeoutEvent, etc. It is necessary for global alert instances to choose event types. However, for normal alert instances, the generation of events is manually triggered by the user (select the alarm strategy and the alarm group bound to the workflow), the types of events generated (ProcessEndEvent, ProcessTimeoutEvent, and ProcessFailEvent) are also fixed. Users have already explicitly bound the alert group and normal alert instances, and the event types which normal alert instance can subscribe to are also fixed, it is fine to send the message directly througn these normal instances. Is it still necessary for users to choose the warning type for normal alert instances?
3. `listener_event` and `alert`: alert content can be abstracted as listener events, and the content stored in `t_ds_listener_event` is actually alert which global alert group need to process.  In this submission, we just add limited logic of global alert events, which is relatively independent with alert.  It’s difficult to merge `listener_event` and `alert` in this submission because `t_ds_alert` contains `process_definition_code`, `project_code` and other information besides the event itself, And the process of generating alerts in `WorkflowExecuteRunnable` and `ProcessAlertManager` is relatively complex. Afterwards, we can consider merging the two.

songjianet commented on dolphinscheduler-ui/src/views/security/alarm-instance-manage/detail.tsx:
```diff
@@ -95,6 +96,16 @@ const DetailModal = defineComponent({
+          warningTypeSpan.value = 0
+        }else{
+          warningTypeSpan.value = 24
+        }
```
warningTypeSpan.value = state.detailForm.instanceType === 'GLOBAL' ? 0 : 24

SbloodyS commented on dolphinscheduler-dao/src/main/resources/sql/dolphinscheduler_mysql.sql:
```diff
@@ -1051,6 +1051,8 @@ INSERT IGNORE INTO `t_ds_version` VALUES ('1', '3.3.0');
 -- ----------------------------
 INSERT IGNORE INTO `t_ds_alertgroup`(alert_instance_ids, create_user_id, group_name, description, create_time, update_time)
 VALUES (NULL, 1, 'default admin warning group', 'default admin warning group', current_timestamp, current_timestamp);
+INSERT IGNORE INTO `t_ds_alertgroup`(alert_instance_ids, create_user_id, group_name, description, create_time, update_time)
+VALUES (NULL, 1, 'global alert group', 'global alert group', current_timestamp, current_timestamp);
```
Please add this to h2.

SbloodyS commented on dolphinscheduler-dao/src/main/resources/sql/dolphinscheduler_postgresql.sql:
```diff
@@ -1071,6 +1073,8 @@ CREATE TABLE t_ds_alert_plugin_instance (
 	create_time timestamp NULL,
 	update_time timestamp NULL,
 	instance_name varchar(255) NULL,
+	instance_type int4 NOT NULL default '0',
```
Please keep the same datatype in all datasouce.

weixiaonan1 committed:
1. sql: sync ddl
2. front-end: change to ternary expression
3. back-end: correct license header in ListenerEvent.java

weixiaonan1 committed:
test case

weixiaonan1 committed:
frontend: remove unnecessary console

weixiaonan1 committed:
fix unit test

weixiaonan1 committed:
Merge branch 'dev' into dev_14832

weixiaonan1 committed:
remove log depends on user-provided value

weixiaonan1 committed:
fix dolphinscheduler_postgresql.sql

weixiaonan1 committed:
sync database schema

weixiaonan1 committed:
fix unit test

weixiaonan1 committed:
fix unit test

weixiaonan1 committed:
Merge branch 'dev' into dev_14832

SbloodyS commented on dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/enums/ListenerEventType.java:
```diff
@@ -0,0 +1,66 @@
+    public static ListenerEventType of(int code) {
+        ListenerEventType listenerEventType = CODE_MAP.get(code);
+        if (listenerEventType == null) {
+            throw new IllegalArgumentException(String.format("The task execution status code: %s is invalidated",
```
```suggestion
            throw new IllegalArgumentException(String.format("The task execution status code: %s is invalid",
```

SbloodyS commented on dolphinscheduler-alert/dolphinscheduler-alert-server/src/main/java/org/apache/dolphinscheduler/alert/service/ListenerEventPostService.java:
```diff
@@ -0,0 +1,262 @@
+@Slf4j
+public final class ListenerEventPostService extends BaseDaemonThread implements AutoCloseable {
+
+    private static final int QUERY_ALERT_THRESHOLD = 100;
```
Can we make this param configurable in config file so that can adjust this value without recompiling.

SbloodyS commented on dolphinscheduler-alert/dolphinscheduler-alert-server/src/main/java/org/apache/dolphinscheduler/alert/service/ListenerEventPostService.java:
```diff
@@ -0,0 +1,262 @@
+                List<ListenerEvent> listenerEvents = listenerEventMapper
+                        .listingListenerEventByStatus(AlertStatus.WAIT_EXECUTION, QUERY_ALERT_THRESHOLD);
+                if (CollectionUtils.isEmpty(listenerEvents)) {
+                    log.debug("There is not waiting listener events");
```
```suggestion
                    log.debug("There is no waiting listener events");
```

SbloodyS commented on dolphinscheduler-alert/dolphinscheduler-alert-server/src/main/java/org/apache/dolphinscheduler/alert/service/ListenerEventPostService.java:
```diff
@@ -0,0 +1,262 @@
+            }
+            AbstractListenerEvent event = generateEventFromContent(listenerEvent);
+            if (event == null) {
+                log.error("parse listener event to abstract listener event fail. {}", listenerEvent.getContent());
```
```suggestion
                log.error("parse listener event to abstract listener event failed: {}", listenerEvent.getContent());
```

SbloodyS commented on dolphinscheduler-dao/src/main/resources/sql/dolphinscheduler_mysql.sql:
```diff
@@ -1086,6 +1088,8 @@ CREATE TABLE `t_ds_alert_plugin_instance` (
   `update_time` timestamp NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
   `instance_name` varchar(255) DEFAULT NULL COMMENT 'alert instance name',
+  `instance_type` int NOT NULL default '0',
+  `warning_type` int,
```
```suggestion
  `warning_type` int NOT NULL,
```

SbloodyS commented on dolphinscheduler-dao/src/main/resources/sql/dolphinscheduler_mysql.sql:
```diff
@@ -2114,3 +2118,21 @@ CREATE TABLE `t_ds_relation_sub_workflow` (
+   `id` int(11) NOT NULL AUTO_INCREMENT COMMENT 'key',
+   `content` text COMMENT 'listener event json content',
+   `sign` char(40) NOT NULL DEFAULT '' COMMENT 'sign=sha1(content)',
+   `post_status` tinyint(4) DEFAULT '0' COMMENT '0:wait running,1:success,2:failed,3:partial success',
+   `event_type` int(11)  COMMENT 'listener event type',
```
```suggestion
   `post_status` tinyint(4) NOT NULL DEFAULT '0' COMMENT '0:wait running,1:success,2:failed,3:partial success',
   `event_type` int(11) NOT NULL  COMMENT 'listener event type',
```

SbloodyS commented on dolphinscheduler-dao/src/main/resources/sql/dolphinscheduler_postgresql.sql:
```diff
@@ -2103,3 +2107,23 @@ CREATE INDEX idx_parent_workflow_instance_id ON t_ds_relation_sub_workflow (pare
+    id          int         NOT NULL,
+    content     text,
+    sign        varchar(40) NOT NULL DEFAULT '',
+    post_status int                  DEFAULT '0',
+    event_type  int,
```
```suggestion
    post_status int  NOT NULL                DEFAULT '0',
    event_type  int NOT NULL,
```

SbloodyS commented on dolphinscheduler-dao/src/main/resources/sql/upgrade/3.3.0_schema/mysql/dolphinscheduler_ddl.sql:
```diff
@@ -14,3 +14,23 @@
+    `id`          int      NOT NULL AUTO_INCREMENT COMMENT "key",
+    `content`     text NULL COMMENT "listener event json content",
+    `sign`        char(40) NOT NULL DEFAULT "" COMMENT "sign=sha1(content)",
+    `post_status` tinyint NULL DEFAULT 0 COMMENT "0:wait running,1:success,2:failed,3:partial success",
+    `event_type`  int NULL COMMENT "listener event type",
```
```suggestion
    `post_status` tinyint NOT NULL DEFAULT 0 COMMENT "0:wait running,1:success,2:failed,3:partial success",
    `event_type`  int NOT NULL COMMENT "listener event type",
```

SbloodyS reviewed (commented):
Generally LGTM. Just some NIT.

ruanwenjun commented on dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/AlertPluginInstanceServiceImpl.java:
```diff
@@ -108,6 +115,20 @@ public Map<String, Object> create(User loginUser, int pluginDefineId, String ins
             log.info("Create alert plugin instance complete, name:{}", alertPluginInstance.getInstanceName());
+            // global instance will be added into global alert group automatically
+            if (instanceType == AlertPluginInstanceType.GLOBAL) {
+                AlertGroup globalAlertGroup = alertGroupMapper.selectById(2);
```
Please don't use magic number 2.

weixiaonan1 replied:
extract the magic number 2 into local variable: GLOBAL_ALERT_GROUP_ID.

ruanwenjun commented on dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ProcessDefinitionServiceImpl.java:
```diff
@@ -306,7 +310,13 @@ public Map<String, Object> createProcessDefinition(User loginUser,
-        return createDagDefine(loginUser, taskRelationList, processDefinition, taskDefinitionLogs, otherParamsJson);
+        result = createDagDefine(loginUser, taskRelationList, processDefinition, taskDefinitionLogs, otherParamsJson);
+        if (result.get(Constants.STATUS) == Status.SUCCESS) {
+            listenerEventAlertManager.publishProcessDefinitionCreatedListenerEvent(loginUser, processDefinition,
+                    taskDefinitionLogs,
+                    taskRelationList);
```
Can we put this in audit aspect?

weixiaonan1 replied:
Do you mean to add annotation on the methods where generate events to make publish events unified?  I think it's hard. Because we cannot get enougn information from arguments and return value of the method for most events such as ProcessStartListenerEvent, ProcessDefinitionDeletedListenerEvent.

weixiaonan1 committed:
fix some NIT.

weixiaonan1 committed:
Merge branch 'dev' into dev_14832

weixiaonan1 committed:
extract GLOBAL_ALERT_GROUP_ID into variable

weixiaonan1 committed:
fix ddl bug

weixiaonan1 committed:
add column task_type in t_ds_fav_task in upgrade/3.2.0_schema

weixiaonan1 committed:
add unit test

zixi0825 commented:
LGTM

SbloodyS reviewed (approved):
+1
